agent/queryparser/helpers.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2019 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package queryparser provides functionality for queries parsing.
+package queryparser
+
+import (
+	"bufio"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+)
+
+var (
+	spaceRegexp *regexp.Regexp
+	spaceOnce   sync.Once
+	errSpace    error
+
+	multilineRegexp *regexp.Regexp
+	multilineOnce   sync.Once
+	errMultiline    error
+)
+
+func parseMySQLComments(q string) (map[string]bool, error) {
+	prepareMultilineRegexp()
+	if errMultiline != nil {
+		return nil, errMultiline
+	}
+
+	// comments using comment as a key to avoid duplicates
+	comments := make(map[string]bool)
+	for _, v := range multilineRegexp.FindAllStringSubmatch(q, -1) {
+		if len(v) < 2 {
+			continue
+		}
+
+		value := removeFormatting(v[1])
+
+		var err error
+		value, err = removeSpaces(value)
+		if err != nil {
+			return nil, err
+		}
+
+		comments[value] = true
+	}
+
+	hashComments, err := parseSinglelineComments(q, "#")
+	if err != nil {
+		return nil, err
+	}
+	for c := range hashComments {
+		comments[c] = true
+	}
+
+	dashComments, err := parseSinglelineComments(q, "--")
+	if err != nil {
+		return nil, err
+	}
+	for c := range dashComments {
+		comments[c] = true
+	}
+
+	return comments, nil
+}
+
+func parsePostgreSQLComments(q string) (map[string]bool, error) {
+	prepareMultilineRegexp()
+	if errMultiline != nil {
+		return nil, errMultiline
+	}
+
+	// comments using comment as a key to avoid duplicates
+	comments := make(map[string]bool)
+	for _, v := range multilineRegexp.FindAllStringSubmatch(q, -1) {
+		if len(v) < 2 {
+			continue
+		}
+
+		value := removeFormatting(v[1])
+
+		var err error
+		value, err = removeSpaces(strings.ReplaceAll(value, "*", ""))
+		if err != nil {
+			return nil, err
+		}
+
+		comments[value] = true
+	}
+
+	dashComments, err := parseSinglelineComments(q, "--")
+	if err != nil {
+		return nil, err
+	}
+	for c := range dashComments {
+		comments[c] = true
+	}
+
+	return comments, nil
+}
+
+func parseSinglelineComments(q, startChar string) (map[string]bool, error) {
+	r, err := regexp.Compile(fmt.Sprintf("%s.*", startChar))
+	if err != nil {
```

Comment:
IMHO don't see much reason checking regexp compile errors on runtime. Maybe 
```
var compiled = regexpr.MustCompite(payload) 
```
at package level would be better? I wonder what is the probability of an error appearing there? )
Here performance [example](https://github.com/golang/go/issues/26623)
```
cat f_test.go
package p

import (
        "regexp"
        "testing"
)

var Sink bool

func BenchmarkCompileRun(b *testing.B) {
        for i := 0; i < b.N; i++ {
                rx := regexp.MustCompile(`[\w\.+-]+@[\w\.-]+\.[\w\.-]+`)
                Sink = rx.MatchString("123456789 foo@bar.etc")
        }
}

func BenchmarkRun(b *testing.B) {
        rx := regexp.MustCompile(`[\w\.+-]+@[\w\.-]+\.[\w\.-]+`)
        for i := 0; i < b.N; i++ {
                Sink = rx.MatchString("123456789 foo@bar.etc")
        }
}
$ go test -bench=.
goos: linux
goarch: amd64
pkg: mvdan.cc/p
BenchmarkCompileRun-4             100000             14160 ns/op
BenchmarkRun-4                   1000000              1121 ns/op
PASS
ok      mvdan.cc/p      2.693s
```